### ISSUE TITLE
Time-less dates

### ISSUE THREAD

**Issue by Philipp91** [dmtrKovalenko/material-ui-pickers#517]
For my use case, users can choose to have a time part on their datetime values or not. They record when certain events happened, and in some cases they only know the date, whereas in other cases they also know the time. (Extensions of this concept would be month granularity, e.g. 2018-07.)

Currently, I end up duplicating quite a bit of the `DateTimePicker` code. First, I need to duplicate the `DateTimePickerHeader` in order to display e.g. "Jul 17   --:--" when there's no time component. Second, I need to duplicate the two outer components because they refer to the wrong, in-library `DateTimePickerHeader`, plus there needs to be a little bit of additional logic and UI.
This ~400 line code duplicate prevents me from merging upstream updates of material-ui-pickers regularly, and it prevents me from contributing fixes and improvements without quite some extra effort.

Proposal A:
If this sounds like a use case that other users might have, I propose to build support for time-less dates into the `utils`, e.g. in form of a `supportsDatesWithoutTime() -> boolean` that enables the feature, `isDateWithoutTime(Date) -> boolean` to recognize such values, and `toDateWithoutTime(Date) -> Date` to eliminate the time part of a `Date`. If supported by their utils, developers could set `<DateTimePicker allowTimeless={true}/>` to show some form of delete button on the time tab, as well as "--:--" in the header if the time part is missing.

date-fns, moment.js and luxon don't seem to support time-less dates in a rigorous way as, e.g., js-joda does: https://js-joda.github.io/js-joda/cheat-sheet.html#localdate
So the default `utils` implementations for those could probably not offer that functionality.
(That said, I personally use a little hack where UTC midnight essentially signals that there is no time information, and I could build that hack into a custom `DateFnUtils` implementation, as could everyone else who doesn't want to use js-joda just for this purpose.)

Proposal B:
If the overhead of maintaining this seems to high given that only few developers would use it (?), I suggest designing the `DateTimePicker` etc. to be (even) more extensible. For instance, they could accept `HeaderProps` and `HeaderComponent` to let callers override the props or even the entire implementation of sub-components if necesssary.

Let me know what you think.

**Comment by dmtrKovalenko** [comment]
But why do you need to remove time from DateTmePicker, if there is a DatePicker itself?

**Comment by Philipp91** [comment]
I believe you are missing the point. Probably my post was too long and confusing, sorry about that.

There's the DatePicker, and I'm using it elsewhere in my app. But this use case is different: Sometimes there is a date, and sometimes there is not -- and the users decide. I guess I could put a `<Switch/>`labeled "has time" and based on that render a `<DatePicker/>` or a `<DateTimePicker/>`, but that would be a horrible UX and consume quite much space in a larger form with many of these pickers.

When I have a time-less date, let's say 2018-07-18, and I open a `<DatePicker/>`, then there can never be any time part. When I open my desired (and currently hacked) `<DateTimePicker/>`, it displays the date, plus a "--:--" in the header. When the user clicks on that, the time becomes "00:00" and the user can select a time, so within one click the date is not time-less anymore.

**Comment by cherniavskii** [comment]
You can use composition to reach the desired result.

- Create a modal.
- Then use `DatePicker` or `DateTimePicker` (in a Switch, as you described above) without a modal wrapper - like this https://codesandbox.io/s/m5n7jjzx1x.

Does this solve your issue?

**Comment by Philipp91** [comment]
Well, it's getting there. See also my proposal B.

In case you care about details, though you probably shouldn't:
I need the tabs always visible, so when there's no time part currently, the user can click the time tab to add it. Thus I need to incorporate the `DateTimePickerTabs` component. Also, I need the header to display "--:--". This may be achievable by overriding the `DateFnUtils`, or by modifying the `DateTimePickerHeader`. And I would need to implement my own text field and modal wrapper even though there's already code for that in material-ui-pickers.

tl;dr: I need to make quite many, tiny adjustments to various components.
However, the library currently isn't very flexible regarding such modifications. Assuming that other people have different, but similarly detailed use cases, a more flexible API would be helpful. To give an artificial example: Assume a developer using the `DateTimePicker` has a requirement to make the `<Clock/>` change its value when Ctrl+scroll-wheel is used. On a `<Clock/> alone, this can be achieved quite easily, but because it's nested very far inside the `<DateTimePicker/>`, the developer probably also has to override all the containing components up to and including the `DateTimePicker` class.
